### PR TITLE
Faster convert for uv_write, no huge boiler-plate assembly, and not throw InexactError potentially while printing

### PR DESCRIPTION
**Pull request by PallHaraldsson**
Using `% Int`, unproblematic/faster on 64-bit, works up to 2 GB strings only on 32-bit, but before this change printed larger strings but then then throws inexact, so kind of didn't work anyway.

### CODE REVIEW

File: base/stream.jl [hunk 1]
```diff
@@ -1100,7 +1100,7 @@ function uv_write(s::LibuvStream, p::Ptr{UInt8}, n::UInt)
     if status < 0
         throw(_UVError("write", status))
     end
-    return Int(n)
+    return n % Int64
 end
```
Comment (giordano) [inline base/stream.jl:1104]: Doesn't this change the return type on 32-bit systems?

Comment (PallHaraldsson) [reply base/stream.jl:1104]: Yes, like I said I thought ok, intentional, but if not I can fix that. Do you really rely on in Int32?

**Comment by KristofferC** [comment]
Typically with performance PRs one also provide a benchmark to show that the PR has any effect in practice. Could you do that?

**Comment by PallHaraldsson** [comment]
I find it very likely, but unsure, this CI error is not related to my change:
```
stream.jl
error during bootstrap:
LoadError("sysimg.jl", 3, LoadError("Base.jl", 450, LoadError("stream.jl", 1072, ErrorException("syntax: cannot juxtapose string literal"))))
jl_f_throw at /cache/build/builder-amdci5-7/julialang/julia-master/src/builtins.c:579
jl_apply at /cache/build/builder-amdci5-7/julialang/julia-master/src/julia.h:2237 [inlined]
do_call at /cache/build/builder-amdci5-7/julialang/julia-master/src/interpreter.c:125
eval_value at /cache/build/builder-amdci5-7/julialang/julia-master/src/interpreter.c:222
eval_stmt_value at /cache/build/builder-amdci5-7/julialang/julia-master/src/interpreter.c:173 [inlined]
eval_body at /cache/build/builder-amdci5-7/julialang/julia-master/src/interpreter.c:685
jl_interpret_toplevel_thunk at /cache/build/builder-amdci5-7/julialang/julia-master/src/interpreter.c:892
top-level scope at stream.jl:1072
jl_toplevel_eval_flex at /cache/build/builder-amdci5-7/julialang/julia-master/src/toplevel.c:1055
jl_eval_throw at /cache/build/builder-amdci5-7/julialang/julia-master/src/toplevel.c:722
jl_eval_errorf at /cache/build/builder-amdci5-7/julialang/julia-master/src/toplevel.c:734
jl_toplevel_eval_flex at /cache/build/builder-amdci5-7/julialang/julia-master/src/toplevel.c:1005
jl_parse_eval_all at /cache/build/builder-amdci5-7/julialang/julia-master/src/toplevel.c:1182
ijl_load_ at /cache/build/builder-amdci5-7/julialang/julia-master/src/toplevel.c:1229
include at ./boot.jl:457
include at ./Base.jl:24
include at ./Base.jl:26
unknown function (ip: 0x7fc115b51752) at (unknown file)
jl_apply at /cache/build/builder-amdci5-7/julialang/julia-master/src/julia.h:2237 [inlined]
do_call at /cache/build/builder-amdci5-7/julialang/julia-master/src/interpreter.c:125
eval_value at /cache/build/builder-amdci5-7/julialang/julia-master/src/interpreter.c:222
eval_stmt_value at /cache/build/builder-amdci5-7/julialang/julia-master/src/interpreter.c:173 [inlined]
eval_body at /cache/build/builder-amdci5-7/julialang/julia-master/src/interpreter.c:685
jl_interpret_toplevel_thunk at /cache/build/builder-amdci5-7/julialang/julia-master/src/interpreter.c:892
top-level scope at Base.jl:450
jl_toplevel_eval_flex at /cache/build/builder-amdci5-7/julialang/julia-master/src/toplevel.c:1055
jl_eval_module_expr at /cache/build/builder-amdci5-7/julialang/julia-master/src/toplevel.c:231 [inlined]
jl_toplevel_eval_flex at /cache/build/builder-amdci5-7/julialang/julia-master/src/toplevel.c:857
jl_parse_eval_all at /cache/build/builder-amdci5-7/julialang/julia-master/src/toplevel.c:1182
ijl_load_ at /cache/build/builder-amdci5-7/julialang/julia-master/src/toplevel.c:1229
include at ./boot.jl:457
unknown function (ip: 0x7fc115b42036) at (unknown file)
jl_apply at /cache/build/builder-amdci5-7/julialang/julia-master/src/julia.h:2237 [inlined]
do_call at /cache/build/builder-amdci5-7/julialang/julia-master/src/interpreter.c:125
eval_value at /cache/build/builder-amdci5-7/julialang/julia-master/src/interpreter.c:222
eval_stmt_value at /cache/build/builder-amdci5-7/julialang/julia-master/src/interpreter.c:173 [inlined]
eval_body at /cache/build/builder-amdci5-7/julialang/julia-master/src/interpreter.c:685
jl_interpret_toplevel_thunk at /cache/build/builder-amdci5-7/julialang/julia-master/src/interpreter.c:892
top-level scope at sysimg.jl:3
jl_toplevel_eval_flex at /cache/build/builder-amdci5-7/julialang/julia-master/src/toplevel.c:1055
jl_parse_eval_all at /cache/build/builder-amdci5-7/julialang/julia-master/src/toplevel.c:1182
ijl_load_ at /cache/build/builder-amdci5-7/julialang/julia-master/src/toplevel.c:1229
ijl_load at /cache/build/builder-amdci5-7/julialang/julia-master/src/toplevel.c:1242
exec_program at /cache/build/builder-amdci5-7/julialang/julia-master/src/jlapi.c:858
true_main at /cache/build/builder-amdci5-7/julialang/julia-master/src/jlapi.c:915
jl_repl_entrypoint at /cache/build/builder-amdci5-7/julialang/julia-master/src/jlapi.c:1059
main at /cache/build/builder-amdci5-7/julialang/julia-master/cli/loader_exe.c:58
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x4010b8) at /workspace/srcdir/glibc-2.17/csu/../sysdeps/x86_64/start.S
*** This error might be fixed by running `make clean`. If the error persists, try `make cleanall`. ***
make[1]: *** [sysimage.mk:70: /cache/build/builder-amdci5-7/julialang/julia-master/usr/lib/julia/sys.ji] Error 1
make[1]: Leaving directory '/cache/build/builder-amdci5-7/julialang/julia-master'
make: *** [Makefile:107: julia-sysimg-ji] Error 2
🚨 Error: The command exited with status 2
```

File: base/stream.jl [hunk 2]
```diff
@@ -1063,10 +1063,13 @@ function readuntil(x::LibuvStream, c::UInt8; keep::Bool=false)
     return bytes
 end
 
-uv_write(s::LibuvStream, p::Vector{UInt8}) = GC.@preserve p uv_write(s, pointer(p), UInt(sizeof(p)))
+uv_write(s::LibuvStream, p::Vector{UInt8}) = GC.@preserve p uv_write(s, pointer(p), sizeof(p) % UInt64) # UInt64 is safe here, also for 32-bit, then get UInt32 back from uv_write
 
 # caller must have acquired the iolock
 function uv_write(s::LibuvStream, p::Ptr{UInt8}, n::UInt)
+    if Int == Int32 && n isa UInt32 && reinterpret(Int32, i) < 0 # aka n >= typemax(Int32) on 32-bit, no-op on 64-bit
```
Comment (Zentrik) [inline base/stream.jl:1070]: `i` is not defined?

File: base/stream.jl [hunk 3]
```diff
@@ -1063,10 +1063,13 @@ function readuntil(x::LibuvStream, c::UInt8; keep::Bool=false)
     return bytes
 end
 
-uv_write(s::LibuvStream, p::Vector{UInt8}) = GC.@preserve p uv_write(s, pointer(p), UInt(sizeof(p)))
+uv_write(s::LibuvStream, p::Vector{UInt8}) = GC.@preserve p uv_write(s, pointer(p), sizeof(p) % UInt64) # UInt64 is safe here, also for 32-bit, then get UInt32 back from uv_write
```
Comment (fingolfin) [inline base/stream.jl:1066]: The function you are calling expects a `UInt`. As such I don't see how converting to a `UInt64` here will do any good, and in particular, in how far it would even be "faster"?!

File: base/stream.jl (hunk 2 above)
Comment (fingolfin) [inline base/stream.jl:1070]: Presumably `n` is meant, but in any case, I fail to see how this code is beneficial. It also seems unrelated to the PR description, it certainly won't make anything "faster"

File: base/stream.jl [hunk 4]
```diff
@@ -1100,7 +1103,7 @@ function uv_write(s::LibuvStream, p::Ptr{UInt8}, n::UInt)
     if status < 0
         throw(_UVError("write", status))
     end
-    return Int(n)
+    return n % Int  # Safe (on 32-bit too, because of check at the top)
```
Comment (fingolfin) [inline base/stream.jl:1106]: So if `n` was larger than `typemax(Int)` (i.e. more than 2 GB), the old code would throw an exception, the new code instead returns a negative value. Based on your PR description (but not the title), is that perhaps you true goal? To not have the code throw here?

**Comment by fingolfin** [comment]
@PallHaraldsson it is very difficult to follow your explanations as to what motivates this PR. In the description you talk about something being "faster" but I see no evidence or at least plausible explanation what should be faster here and why.

On the other than it seems this PR is about writing more than 2GB at a time on 32bit systems. The change you make seems to be to have it throw an exception *at the start* and not at the end, with a maybe more helpful error message?

What is it now?

File: base/stream.jl [hunk 5]
```diff
@@ -1063,10 +1063,13 @@ function readuntil(x::LibuvStream, c::UInt8; keep::Bool=false)
     return bytes
 end
 
-uv_write(s::LibuvStream, p::Vector{UInt8}) = GC.@preserve p uv_write(s, pointer(p), UInt(sizeof(p)))
+uv_write(s::LibuvStream, p::Vector{UInt8}) = GC.@preserve p uv_write(s, pointer(p), sizeof(p) % UInt64) # UInt64 is safe here, also for 32-bit, then get UInt32 back from uv_write
 
 # caller must have acquired the iolock
 function uv_write(s::LibuvStream, p::Ptr{UInt8}, n::UInt)
+    if Int == Int32 && n isa UInt32 && reinterpret(Int32, i) < 0 # aka n >= typemax(Int32) on 32-bit, no-op on 64-bit
+        throw(ArgumentError(LazyString("On 32-bit strings larger than 2 GB can't be printed in one go! Iterating over strings and printing the Chars is always safe, but splitting string into two SubStrings and printing sepratelyly may not be.")))
```
Comment (fingolfin) [inline base/stream.jl:1071]: This error text is problematic. Note that we are inside `uv_write` which is not just for *printing strings* but writing out memory buffers in a more general setting. There are also typos in there.

I am not yet convinced we want this exception, but if we do, maybe the message should be more on point, e.g.
```suggestion
        throw(ArgumentError(LazyString("cannot write more than 2 GB at a time")))
```
or so

File: base/stream.jl (hunk 3 above)
Comment (PallHaraldsson) [reply base/stream.jl:1066]: UInt is UInt64 on 64-bit, so ok. UInt is Int32 on 32-bit why I comment that this is still ok.

File: base/stream.jl (hunk 2 above)
Comment (PallHaraldsson) [reply base/stream.jl]: Yes. fixed.

File: base/stream.jl (hunk 4 above)
Comment (PallHaraldsson) [reply base/stream.jl]: Note two cases a) 32-bit and b) 64-bit:

a) No it should not return a negative value because of the check at the top making it impossible.

b) On 64-bit strings of length 2^63 (and larger, i.e. "impossible" lengths) would have returned negative with this change, but Inexact as it was, and pay for that check if it passes or not.

**Comment by PallHaraldsson** [comment]
It was a one-line change in the beginning. I saw in-needed code (on 64-bit). with `@code_native`

I thought about effect on 32-bit, but then made changes to conform to the original return type (for 32-bit platforms), why the code is a bit non-obvious... after my further change.

I'm not very concerned making 32-bit faster, but I think it's at least not slower, and with such systems often having less (L1) cache likely more helpful there.

**Comment by PallHaraldsson** [comment]
>The change you make seems to be to have it throw an exception at the start and not at the end, with a maybe more helpful error message?

Exactly (but wasn't the original point or then in the code).

You can construct string larger than 2GB on 32-bit, AND then print them (I didn't test), and THEN you would get InExact error, so it's not hugely important to be able to print first.

It's of course a very obscure error to throw (and after the fact) and if you try to catch the exception and continue, then it seems like a bug, at least a memory leak if it would actually work.

I kind of liked just returning UInt64 also on 32-bit, would have been simpler code. :) And no throwing or need to explain workaround.

File: base/stream.jl (hunk 3 above)
Comment (PallHaraldsson) [reply base/stream.jl:1066]: I see this has a bug on 32-bit after all, will fix.

**Comment by fingolfin** [comment]
I do not believe this has merit.
